src/EditorFeatures/CSharpTest/CompleteStatement/CSharpCompleteStatementCommandHandlerTests.cs

```diff
@@ -0,0 +1,1296 @@
+        public void DontComplete_MethodNameOfMethodInvocation1()
+        {
+            var code = CreateTestWithMethodCall(@"var test = ClassC.Meth$$odM(x,y);");
+
+            var expected = CreateTestWithMethodCall(@"var test = ClassC.Meth;$$odM(x,y);");
+
+            VerifyTypingSemicolon(code, expected);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CompleteStatement)]
+        public void DontComplete_MethodNameOfMethodInvocation2()
+        {
+            var code = CreateTestWithMethodCall(@"var test = ClassC.$$MethodM(x,y);");
+
+            var expected = CreateTestWithMethodCall(@"var test = ClassC.;$$MethodM(x,y);");
+
+            VerifyTypingSemicolon(code, expected);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CompleteStatement)]
+        public void DontComplete_MethodNameOfMethodInvocation3()
+        {
+            var code = CreateTestWithMethodCall(@"var test = ClassC.MethodM$$(x,y);");
+
+            var expected = CreateTestWithMethodCall(@"var test = ClassC.MethodM;$$(x,y);");
+
+            VerifyTypingSemicolon(code, expected);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CompleteStatement)]
+        public void DontComplete_SemicolonBeforeEquals()
+        {
+            var code = CreateTestWithMethodCall(@"var test $$= ClassC.MethodM(x,y);");
+
+            var expected = CreateTestWithMethodCall(@"var test ;$$= ClassC.MethodM(x,y);");
+
+            VerifyTypingSemicolon(code, expected);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CompleteStatement)]
+        public void DontComplete_SemicolonAfterEquals()
+        {
+            var code = CreateTestWithMethodCall(@"var test =$$ ClassC.MethodM(x,y);");
+
+            var expected = CreateTestWithMethodCall(@"var test =;$$ ClassC.MethodM(x,y);");
+
+            VerifyTypingSemicolon(code, expected);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CompleteStatement)]
+        public void ArgumentListOfMethodInvocation1_MissingParen()
+        {
+            var code = CreateTestWithMethodCall(@"var test = ClassC.MethodM($$x, y");
+
+            var expected = CreateTestWithMethodCall(@"var test = ClassC.MethodM(x, y);$$");
+
+            VerifyTypingSemicolon(code, expected);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CompleteStatement)]
+        public void ArgumentListOfMethodInvocation2_MissingParen()
+        {
+            var code = CreateTestWithMethodCall(@"var test = ClassC.MethodM(x$$, y");
+
+            var expected = CreateTestWithMethodCall(@"var test = ClassC.MethodM(x, y);$$");
+
+            VerifyTypingSemicolon(code, expected);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.CompleteStatement)]
+        public void ArgumentListOfMethodInvocation3_MissingParen()
+        {
+            var code = CreateTestWithMethodCall(@"var test = ClassC.MethodM(x,$$ y");
```

Comment:
does this mean, it is physically impossible to put ";" at the $$ location? if user do "undo", will ";" moved to $$ place?

Reply:
I'm not sure we need an intermediate state for the undo operation. Statement completion on semicolon only modifies the behavior in a limited subset of cases where the semicolon couldn't be valid at the current location (or at least that's the intent).